### ISSUE TITLE
Windowed aggregations: send only the final aggregation result of each window (cf. Suppress API of KStreams)

### ISSUE THREAD

**Issue by ybyzek** [confluentinc/ksql#1030]
tl;dr : would like a stream that captures only the final aggregation result of each window.  It is ok to disregard late-arriving data (or define wait period before disregarding).

Looking to achieve “send only the final aggregation result of each window”, as described here: https://stackoverflow.com/questions/38935904/how-to-send-final-kafka-streams-aggregation-result-of-a-time-windowed-ktable  .   There is some coded solution available with the Java Kafka Streams API, described http://blog.inovatrend.com/2018/03/making-of-message-gateway-with-kafka.html but it seems we cannot do this in KSQL.

There are some knobs to play with, e.g. cache size (`cache.max.bytes.buffering`) and commit interval (`commit.interval.ms`), but they just change the update interval.  It doesn't quite accomplish the above request.

**Comment by apurvam** [comment]
This seems to be a dup of #778. If so can you please close it @ybyzek ?

**Comment by ybyzek** [comment]
@apurvam , through my lens, the use case may be different: #778 is aggregates upon aggregates, whereas #1030 is simply emitting one result per window even when it's not aggregates upon aggregates.

However, it is possible that the fix is the same :)  If same fix, yes, let's close this one.  But if there is a possibility that the fixes for each would be different...?

**Comment by big-andy-coates** [comment]
This feels more like a KStreams feature request - with KSQL support to follow.

**Comment by rodesai** [comment]
I don't see how this is possible in the strictest sense. KSQL/Kafka Streams has no way of knowing (I guess other than when a windows retention is up) when all records for a given window have been processed, and therefore has no way of knowing when to send the final window aggregate. The best we could do is use some heuristic to guess that everything in the window is processed. But is that much better than using the cache/commit-interval settings?

**Comment by apurvam** [comment]
This is only possible if we do 'micro batching', ie. emit the aggregates once the window retention has passed. This means that there is a fixed latency to the outputs. 

It would be interesting to see the use cases for this specific feature (only one value for the aggregate per window). My suspicion is that there are other ways to realize that use case, or those use cases may be better served by non streaming systems.

**Comment by miguno** [comment]
Kafka Streams is currently working on the foundational feature for this, see [KIP-328: Ability to suppress updates for KTables](https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables).  Once this is available in Kafka Streams, we can discuss how to leverage it to provide the same functionality in KSQL.

**Comment by miguno** [comment]
With the first iteration of [KIP-328: Ability to suppress updates for KTables](https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables) completed, Kafka Streams in upcoming Apache Kafka 2.1 supports "sending a single, final output record for each window".

**Comment by NingLee** [comment]
@miguno Does [KIP-328: Ability to suppress updates for KTables](https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables) is supported now? I find that Kafka 2.1 is released, but KIP-328 is not mentioned in release note.

**Comment by yaojingguo** [comment]
`suppress` has been supported in [Confluent confluent-5.1.0](https://docs.confluent.io/current/release-notes.html).
> Sending a single, final output for windowed operations

**Comment by miguno** [comment]
@NingLee : Suppressing updates is supported in Kafka Streams in Kafka 2.1, yes.  See docs at https://kafka.apache.org/21/documentation/streams/developer-guide/dsl-api.html#window-final-results.

**Comment by pbkdf3** [comment]
This issue was about exposing "one output per window"/"suppression" in ksql not Kafka Streams.  Why was it closed, does ksql now support this?  There is no mention in the above linked release notes and can't find it in the documentation.

**Comment by alex88** [comment]
@pbkdf3 this wasn't closed, as ksql still doesn't support it

**Comment by hvazquezr** [comment]
Hello everyone, I'm in this of this functionality as well for doing clickstream funnel analysis. Does anyone know if this can be expected anytime soon? Thanks!

**Comment by hvazquezr** [comment]
> This is only possible if we do 'micro batching', ie. emit the aggregates once the window retention has passed. This means that there is a fixed latency to the outputs.
> 
> It would be interesting to see the use cases for this specific feature (only one value for the aggregate per window). My suspicion is that there are other ways to realize that use case, or those use cases may be better served by non streaming systems.

In my particular case you can define as an event the fact that a user started a transaction and didn't finish within X amount of time; and ther is a desire to take a quick action about this; for example, providing some sort of stimulus to the user to close the transaction/purchase/enrollment

**Comment by kevin90828** [comment]
what about this questions now?

**Comment by kevin90828** [comment]
how about this question now?

**Comment by kevin90828** [comment]
> Kafka Streams is currently working on the foundational feature for this, see [KIP-328: Ability to suppress updates for KTables](https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables). Once this is available in Kafka Streams, we can discuss how to leverage it to provide the same functionality in KSQL.

now,as ksql still doesn't support it

**Comment by wangshao1** [comment]
Will set commit.interval.ms to be the same with window size help?

**Comment by jeffreylutz** [comment]
The root cause of the issue is actually a property in KSQL in Confluent's docker image not being set.

The docker image from Confluent doesn't set an important property: ksql.streams.cache.max.bytes.buffering

So, set this property for in ksql:  Here is the link to the doc:  https://docs.confluent.io/current/ksql/docs/installation/server-config/config-reference.html#ksql-streams-cache-max-bytes-buffering

For more information on this KSQL property, read this:  https://docs.confluent.io/current/streams/developer-guide/memory-mgmt.html

**Comment by alex3165** [comment]
@big-andy-coates @JimGalasyn @hjafarpour @rodesai Any update on the above? I think sending the last aggregation of a time window is an expected behaviour for must of us using window tumbling.

**Comment by rjmccabe3701** [comment]
@alex3165 I agree, the current behavior is pretty useless for me -- If I'm indicating I want aggregated results every, say 10 minutes, I'm not expecting aggregate results until the window is done.  So ... How do I get the results I want with just ksql?  Is there anyway to sample the aggregate stream just at the window-end?

**Comment by derekjn** [comment]
This is now available in KStreams as of 2.3, so we should be able to support it now.

**Comment by abhi-dnyate** [comment]
Hello,
Any rough idea, when would it be supported for KSQL ?

**Comment by almazik** [comment]
Proposal KLIP has been merged with the Release Target: 5.5. Any ideas of the real timeframe for this feature? We are trying to leverage KSQL to reduce the load on services by aggregating the raw data over reasonable time periods (minutes/hours/days). 

But there is no practical reason for us to use KSQL over manual aggregation without this feature.

**Comment by agavra** [comment]
@almazik I had forgotten to update the release target, if you are using CP releases it will likely be merged in 6.1 - if you are using ksqlDB, then it will likely be in 0.11 or 0.12 depending on when we decide to cut those releases (about 1-2 months time)

**Comment by vpalmisano** [comment]
Hi to all. Is it possible to add an option to the EMIT FINAL statement (or using other options as well) in order to use the `untilTimeLimit` instead of `untilWindowCloses` (the default value now). This will allow the rate limiting of aggregate queries. Thanks

**Comment by agavra** [comment]
@vpalmisano - we had considered doing this, but decided against it because of the confusion that Kafka Streams saw with `untilTimeLimit`. Note that this is _stream time_ and not wall clock time, which makes it a little difficult to reason about. When wall clock time is available, we can reconsider that. Can you submit a new ticket to track this?
